core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -79,6 +88,44 @@ public static CloseableIterable<FileScanTask> splitFiles(
     return CloseableIterable.combine(splitTasks, tasks);
   }
 
+  /**
+   * Produces {@link CombinedScanTask combined tasks} from an iterable of {@link FileScanTask file
+   * tasks}, using an adaptive target split size that targets a minimum number of tasks
+   * (parallelism).
+   *
+   * @param files incoming iterable of file tasks
+   * @param parallelism target minimum number of tasks
+   * @param splitSize target split size
+   * @param lookback bin packing lookback
+   * @param openFileCost minimum file cost
+   * @return an iterable of combined tasks
+   */
+  public static CloseableIterable<CombinedScanTask> planTasksAdaptive(
+      CloseableIterable<FileScanTask> files,
+      int parallelism,
+      long splitSize,
+      int lookback,
+      long openFileCost) {
+
+    validatePlanningArguments(splitSize, lookback, openFileCost);
+
+    Function<FileScanTask, Long> weightFunc =
+        file ->
+            Math.max(
```

Comment:
I'm a little confused by the way we're calculating this weight function.  It looks like we're taking the larger of:
1.  the total file size (including deletes)
2. the total number of files * open cost

Shouldn't the the weight of the task be the total file size plus the open cost for each file?  With 1. above we're not including the open cost and in 2. we're not including the data to be read.

Reply:
This is the way it was before, but I also flagged this as something to look into.

I think we should be doing the `Math.max(fileSize, openCost)` for every file individually. Otherwise we could overpack small delete files because they aren't large compared to the size of data files.

Reply:
I was looking at bin packing and it seems like we would always need similar weights (e.g. you can't use file size for one task and open cost for another because they're too dissimilar).  If open cost is in bytes, it feels like this should just be `totalFileSize + (fileCount * openCost)`.

Reply:
The purpose of the open file cost was to avoid packing lots of tiny KB-sized files into a single split of 128 MB, which would create straggler tasks with many files. When we added this, we only had data files and we thought the cost of opening files should not matter if file sizes are reasonable.

I still tend to believe the cost of opening files should not matter for reasonably sized files to avoid changes in the way splits are packed right now. Comparing `fileSize` and `openCost` per each file seems worth exploring.

Reply:
> Comparing fileSize and openCost per each file seems worth exploring.

This seems more correct to me. Otherwise, a single large file could be combined with lots of small files and delete files that take too long. This is for a separate PR though.